core/server/models/session.js

```diff
@@ -0,0 +1,54 @@
+const ghostBookshelf = require('./base');
+
+let Session,
+    Sessions;
+
+Session = ghostBookshelf.Model.extend({
+    tableName: 'sessions',
+
+    parse: function parse(attrs) {
+        if (!attrs.session_data) {
+            return attrs;
+        }
+        return Object.assign(attrs, {
+            session_data: JSON.parse(attrs.session_data)
+        });
+    },
+
+    format: function format(attrs) {
+        if (!attrs.session_data) {
+            return attrs;
+        }
+        return Object.assign(attrs, {
+            session_data: JSON.stringify(attrs.session_data)
+        });
+    },
+
+    user: function () {
+        return this.belongsTo('User');
+    }
+}, {
+    upsert: function (id, data) {
+        return this.forge({id})
+            .fetch({require: true})
+            .then((model) => {
+                return model.set('session_data', data).save();
+            }, () => {
+                return this.forge({
```

Comment:
The model layer has an inherited structure.

The base model offers all CRUD functions and offers a set of general logic.

`return ghostBookshelf.Model.add.call(this, data, options)`

Same for `edit`.

Reply:
I don't think we want to use these for the `SessionStore` methods, we have quite custom logic, e.g. we choose our own id, and destroy should not error if the session is missing. What do you think?

Reply:
Hmmm not sure. Let's go over your examples.

> destroy should not error if the session is missing

You could call the super method "destroy".
You could catch the response and return success?

> we choose our own id

e.g. `onSaving` generates the model id. But you can also register `onSaving` in the session model and don't call super, because you want to control creating sessions on your own.

___

If we do things like this 

```
models.Session.forge().destroy
```

we call `destroy` and this is method from the instance. It goes straight into bookshelf.

Some more insights:

- the base model layer offers all kinds of hooks: filtering, sanitize hooks etc:
- the base model auto sanitizes stuff like dates
- the base model is transactional/blocking for update actions

The base model layer is defo not perfect, but we want to improve it over time.